Support component cycles in stub generation

westse opened the pull request:
<!--
Thanks for your interest in the project. Bugs filed and PRs submitted are appreciated!

Please make sure that you are familiar with and follow the Code of Conduct for
this project (found in the CODE_OF_CONDUCT.md file).

Also, please make sure you're familiar with and follow the instructions in the
contributing guidelines (found in the CONTRIBUTING.md file).

Please fill out the information below to expedite the review and (hopefully)
merge of your pull request!
-->

<!-- What changes are being made? (What feature/bug is being fixed here?) -->

**What**: Fixes #591 and StackOverflowError produced when generating stubs for spec containing component cycles.

<!-- Why are these changes necessary? -->

**Why**: Avoids serious StackOverflowError and better supports the OpenApi specification

<!-- How were these changes implemented? -->

**How**: Resolver keeps a stack of patterns visited in order to avoid infinite processing cycle.
- Otherwise StackOverflowError thrown
- Uses correct pattern for the referred-to component that would introduce a cycle (previously would use the referring component's pattern). See OpenApiSpecification.toSpecmaticPattern()
- Test enhanced to test stub generation for direct cycle, indirect cycles through another component, and cycles through arrays.

<!-- Have you done all of these things?  -->

**Checklist**:

<!-- add "N/A" to the end of each line that's irrelevant to your changes -->

<!-- to check an item, place an "x" in the box like so: "- [x] Documentation" -->

- [ ] Documentation added to the README.md OR link to PR on https://github.com/specmatic/specmatic-documentation
- [X] Tests
- [ ] Sonar Quality Gate

<!-- Kindly link the issue that this PR will be addressing -->
**Issue ID**:
Closes: #591 

<!-- feel free to add additional comments -->
Related to #583 , but this issue deals with stub generation logic.

westse committed:
Pattern cycle handling
- Avoids StackOverflowError that is otherwise thrown
- Cycles in non-nullable patterns are not valid and therefore produce a ContractException
- Cycles in nullable patterns are valid so are supported. These result in an AnyPattern where NullPattern is one of this possible patterns chosen, so this will guarantee an eventually terminated cycle. The logic that detects cycles allows an exception for cycles involving nullable patterns.
- Uses correct pattern for the referred-to component that would introduce a cycle (previously would use the referring component's pattern). See OpenApiSpecification.toSpecmaticPattern()

harikrishnan83 commented on core/src/main/kotlin/in/specmatic/core/Resolver.kt:
```diff
@@ -70,8 +72,18 @@ data class Resolver(
         }
 
     fun generate(factKey: String, pattern: Pattern): Value {
```
@westse Thanks for covering this in ```OpenApiKtTest.kt```. However, can we also cover this method with a unit test in ```ResolverKtTest```? As you may have noticed already our ```OpenApiKtTest``` is becoming quite large and a little hard to manage all functionality at that level. Also since Resolver is not particularly associated with openapi it may be easier / appropriate to test it independently anyway.

westse replied:
That makes sense - but since you plan to refactor the cyclePreventionStack anyway perhaps you can add it then?

joelrosario reviewed (changes requested):
Hi @westse 

We understand the problem highlighted in the issue #591. Specmatic should certainly not be throwing a StackOverflowException.

However, given that the cyclical reference in your example / test in this PR is not nullable, the resolver cannot return a null to terminate the cycle, as that would not be contract-valid.

Can you update the PR to throw a ContractException exception, indicating the type that was found in the typeStack in the resolver, which has a cyclical reference.

Note: The API Specification can be updated to make the cyclical reference a nullable value. Nullable values are represented by AnyPattern, containing NullPattern as one of the candidate values. When AnyPattern generates a value, it picks randomly between the candidate values. At some point, it would pick NullPattern and generate a value, thus breaking the cycle.

westse committed:
Fix off-by-one condition

westse committed:
Fix pattern cycle on generate
- Allows cycle to be terminated by omitting an optional property that would have introduced the cycle.

harikrishnan83 commented on core/src/main/kotlin/in/specmatic/core/pattern/JSONObjectPattern.kt:
```diff
@@ -102,9 +103,27 @@ data class JSONObjectPattern(override val pattern: Map<String, Pattern> = emptyM
+                try {
+                    Optional.of(resolverWithNullType.generate(key, pattern))
+                } catch (e: PatternCycleException) {
+                    if (optionalProps.contains(key)) {
```
@westse In this location, the optionality of the key itself may be relevant to break the cycle. What we should checking for is if the nullability of the value. We will push this fix to your branch shortly so that you can also take a look at the changes and share your thoughts. Hope that is okay. cc @joelrosario

westse replied:
Sorry, I'm not sure I follow. I may be missing a concept, but let me describe the problem I'm addressing and this current solution. (Also, I apologize that somehow I originally missed including a unit test to expose the limitation being addressed. I have since pushed a test along with a necessary refactor.)

Stepping back, this PR is meant to have specmatic behave well in the face of cycles in patterns (e.g. A->B->A). Per @joelrosario 's feedback, cycles in required, non-nullable patterns should result in an exception. However cycles in nullable patterns (both required and optional) can be allowed because we can rely on AnyPattern.generate()'s behavior to eventually randomly select null and terminate the cycle. (That is what the Resolver cyclePreventionStack was for)

However, there is another case: cycles where one or more non-nullable properties are optional (ie. do no appear in the required list). Rather than failing due to a cycle it seems preferrable (and contract-valid) to instead omit the optional property in the generated stub. That is what this logic is meant to accomplish.

harikrishnan83 commented on core/src/main/kotlin/in/specmatic/core/Resolver.kt:
```diff
@@ -69,9 +70,35 @@ data class Resolver(
+
+            validateNoCycle(pattern)
+
+            cyclePreventionStack.add(pattern)
```
@westse We notice that cyclePreventionStack is a multable list which is an instance variable of Resolver, which thereby makes Resolver itself mutable. Because of the nature of how we pass resolver around, it is extremely critical to keep the resolve immutable. We will make the necessary changes (passing the type stack as method parameter thereby not requiring instance variable) and push it your branch. cc @westse

westse replied:
Agreed about the immutability (I was just trying to not be too invasive :smile: ).

harikrishnan83 reviewed (commented):
@westse Thanks again.

@joelrosario and I have taken a look at this PR and we have left a couple of comments. We need to make two changes to this PR before we can merge it.
* Logical problem in JSONObjectPattern where we are handling PatternCycleException
* Resolver by design has to be immutable. The cyclePreventionStack by virtue of being a mutable collection also makes Resolver mutable, which needs to be addressed.
As mentioned in the comments, we will make these changes and push them to your branch.

westse committed:
Test for cycles with optional props
- Refactored to allow previous patterns in the cycle a chance to omit optional properties that otherwise would have introduced a cycle.

westse committed:
Test for cycles with required nullable props
- To properly test that nullable components don't result in a cycle detection error the components must be referred to using required properties - otherwise the test might artificially pass due to separate (but similar) logic for optional properties that also prevents a cycle detection error.
